ambry-router/src/main/java/com/github/ambry/router/ParanoidDurabilityOperationTracker.java

```diff
@@ -0,0 +1,314 @@
+package com.github.ambry.router;
+
+import com.github.ambry.clustermap.PartitionId;
+import com.github.ambry.clustermap.ReplicaId;
+import com.github.ambry.clustermap.ReplicaState;
+import com.github.ambry.config.RouterConfig;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An implementation of {@link OperationTracker} for PUTs only (it does not support any other request types). It differs
+ * from {@link SimpleOperationTracker} in that it is "paranoid" about durability. In addition to writing the new blob to
+ * the required number of local replicas, it also requires writes to succeed for at least one replica in one remote
+ * data center, effectively circumventing cross-datacenter replication (configurable). This is useful for clients such
+ * as Samza, that require strong durability guarantees.
+ *
+ */
+public class ParanoidDurabilityOperationTracker extends SimpleOperationTracker {
+  private static final Logger logger = LoggerFactory.getLogger(ParanoidDurabilityOperationTracker.class);
+  protected final int localReplicaSuccessTarget;
+  protected int remoteReplicaSuccessTarget = 0;
+  protected final int localReplicaParallelism;
+  protected final int remoteReplicaParallelism;
+  protected int localInflightCount = 0;
+  protected int remoteInflightCount = 0;
+  protected int localReplicaSuccessCount = 0;
+  protected int remoteReplicaSuccessCount = 0;
+  private final Map<String, LinkedList<ReplicaId>> replicaPoolByDc = new HashMap<>();
+  private final ParanoidDurabilityTrackerIterator otIterator;
+  private Iterator<ReplicaId> replicaByDcIterator;
+  private Map<String, Integer> failedCountByDc;
+  private Map<String, Integer> replicaSuccessCountByDc;
+  private Map<String, Integer> disabledCountByDc;
+  private Map<String, Integer> replicaInPoolOrFlightCountByDc;
+  private Map<String, Integer> totalReplicaCountByDc;
+  private Set<String> allDatacenters;
+  private List<String> remoteDatacenters;
+
+  ParanoidDurabilityOperationTracker(RouterConfig routerConfig, PartitionId partitionId, String originatingDcName,
+      NonBlockingRouterMetrics routerMetrics) {
+    super(routerConfig, RouterOperation.PutOperation, partitionId, originatingDcName, true, routerMetrics);
+
+    localReplicaParallelism = routerConfig.routerPutLocalRequestParallelism;
+    remoteReplicaParallelism = routerConfig.routerPutRemoteRequestParallelism;
+    localReplicaSuccessTarget = routerConfig.routerPutLocalSuccessTarget;
+    remoteReplicaSuccessTarget = routerConfig.routerPutRemoteSuccessTarget;
+
+    List<ReplicaId> allEligibleReplicas = getEligibleReplicas(datacenterName, EnumSet.of(ReplicaState.STANDBY, ReplicaState.LEADER));
```

Comment:
To get eligible replicas for all datacenters, we have to pass a "null" parameter to first argument.

Reply:
Good catch, I meant to pass the null but somehow missed it.